TELCODOCS-1826: Day 0 - Day 2 kmods (#75752)
openshift/openshift-docs: modules/kmm-day0-day2-lifecycle-management.adoc

// Module included in the following assemblies:
//
// * updating/preparing_for_updates/kmm-preflight-validation.adoc

:_mod-docs-content-type: CONCEPT
[id="kmm-day0-day2-lifecycle-management_{context}"]
= Lifecycle management

You can leverage KMM to manage the Day 0 through Day 2 lifecycle of kmods without a reboot when the driver allows it.

[NOTE]
====
This will not work if the upgrade requires a node reboot, for example, when rebuilding `initramfs` files is needed.
====

Use one of the following options for lifecycle management.

> .Treat the kmod as an in-tree driver

Author: Changed to headings.

> Use this method when you want to upgrade the kmods. In this case, treat it as an in-tree driver and create a `Module` in the cluster with the `inTreeRemoval` field to unload the old version of the driver.

Author: it=kmod. Changed.

Note the following characteristics of treating the kmod as an in-tree driver:

> * Downtimes could occur, as KMM tries to unload and load the kmod on all the selected nodes simultaneously.

Author: changed.

> * This works in case removing the driver makes the node lose connectivity because KMM uses a single pod to unload and load the driver.

Author: Changed.

.Use ordered upgrade

> In this case, you can use ordered upgrade (ordered_upgrade.md) to create a versioned `Module` in the cluster representing the kmods with no effect because the kmods are already loaded.

Author: Changed.

Note the following characteristics of using ordered upgrade:

> * No cluster downtime because you control the pace of the upgrade and how many nodes are upgraded at the same time, therefore, an upgrade with no downtime is possible.

Author: Changed.

> * This method will not work if unloading the driver results in losing connection to the node, because KMM creates two different worker pods, one for unloading and another for loading, which will not be scheduled.

Reviewer: Needs some editing so it's not a run-on sentence. Also it's not clear what "which will not be scheduled" refers to - pods?

suggestion:

```suggestion
* This method will not work if unloading the driver results in losing connection to the node, because KMM creates two different worker pods for unloading and loading. These pods will not be scheduled.
```

Author: Correct, we are referring to the pods. Made these changes.
